File: db/csharp/SalesManagement.Infrastructure/Repositories/PurchaseOrderDetailRepository.cs
using Dapper;
using Npgsql;
using SalesManagement.Domain.Models;

namespace SalesManagement.Infrastructure.Repositories
{
    /// <summary>
    /// 発注データ明細のRepositoryクラス
    /// </summary>
    public class PurchaseOrderDetailRepository
    {
        private readonly string _connectionString;

        public PurchaseOrderDetailRepository(string connectionString)
        {
            _connectionString = connectionString;
        }

        /// <summary>
        /// 発注明細を登録
        /// </summary>
        public async Task InsertAsync(PurchaseOrderDetail detail)
        {
            const string sql = @"
                INSERT INTO 発注データ明細 (
                    発注番号, 発注行番号, 商品コード, 商品名, 発注単価, 発注数量,
                    消費税率, 入荷済数量, 値引金額, 指定納期,
                    作成日時, 作成者名, 更新日時, 更新者名
                ) VALUES (
                    @PoNo, @PoRowNo, @ProductCode, @ProductName, @UnitPrice, @Quantity,
                    @ConsumptionTaxRate, @ReceivedQuantity, @Discount, @DueDate,
                    @CreatedAt, @CreatedBy, @UpdatedAt, @UpdatedBy
                )";

            await using var connection = new NpgsqlConnection(_connectionString);
            await connection.ExecuteAsync(sql, detail);
        }

        /// <summary>
        /// 発注明細を更新
        /// </summary>
        public async Task UpdateAsync(PurchaseOrderDetail detail)
        {
            const string sql = @"
                UPDATE 発注データ明細
                SET 商品コード = @ProductCode,
                    商品名 = @ProductName,
                    発注単価 = @UnitPrice,
                    発注数量 = @Quantity,
                    消費税率 = @ConsumptionTaxRate,
                    入荷済数量 = @ReceivedQuantity,
                    値引金額 = @Discount,
                    指定納期 = @DueDate,
                    更新日時 = @UpdatedAt,
                    更新者名 = @UpdatedBy
                WHERE 発注番号 = @PoNo
                  AND 発注行番号 = @PoRowNo";

            await using var connection = new NpgsqlConnection(_connectionString);
            await connection.ExecuteAsync(sql, detail);
        }

        /// <summary>
        /// 発注明細を削除
        /// </summary>
        public async Task DeleteAsync(string poNo, int poRowNo)
        {
            const string sql = @"
                DELETE FROM 発注データ明細
                WHERE 発注番号 = @PoNo
                  AND 発注行番号 = @PoRowNo";

            await using var connection = new NpgsqlConnection(_connectionString);
            await connection.ExecuteAsync(sql, new { PoNo = poNo, PoRowNo = poRowNo });
        }

        /// <summary>
        /// 発注番号と行番号で検索
        /// </summary>
        public async Task<PurchaseOrderDetail?> FindByIdAsync(string poNo, int poRowNo)
        {
            const string sql = @"
                SELECT
                    発注番号 AS PoNo,
                    発注行番号 AS PoRowNo,
                    商品コード AS ProductCode,
                    商品名 AS ProductName,
                    発注単価 AS UnitPrice,
                    発注数量 AS Quantity,
                    消費税率 AS ConsumptionTaxRate,
                    入荷済数量 AS ReceivedQuantity,
                    値引金額 AS Discount,
                    指定納期 AS DueDate,
                    作成日時 AS CreatedAt,
                    作成者名 AS CreatedBy,
                    更新日時 AS UpdatedAt,
                    更新者名 AS UpdatedBy
                FROM 発注データ明細
                WHERE 発注番号 = @PoNo
                  AND 発注行番号 = @PoRowNo";

            await using var connection = new NpgsqlConnection(_connectionString);
            return await connection.QueryFirstOrDefaultAsync<PurchaseOrderDetail>(
                sql, new { PoNo = poNo, PoRowNo = poRowNo });
        }

        /// <summary>
        /// 発注番号で検索
        /// </summary>
        public async Task<IEnumerable<PurchaseOrderDetail>> FindByPoNoAsync(string poNo)
        {
            const string sql = @"
                SELECT
                    発注番号 AS PoNo,
                    発注行番号 AS PoRowNo,
                    商品コード AS ProductCode,
                    商品名 AS ProductName,
                    発注単価 AS UnitPrice,
                    発注数量 AS Quantity,
                    消費税率 AS ConsumptionTaxRate,
                    入荷済数量 AS ReceivedQuantity,
                    値引金額 AS Discount,
                    指定納期 AS DueDate,
                    作成日時 AS CreatedAt,
                    作成者名 AS CreatedBy,
                    更新日時 AS UpdatedAt,
                    更新者名 AS UpdatedBy
                FROM 発注データ明細
                WHERE 発注番号 = @PoNo
                ORDER BY 発注行番号";

            await using var connection = new NpgsqlConnection(_connectionString);
            return await connection.QueryAsync<PurchaseOrderDetail>(sql, new { PoNo = poNo });
        }

        /// <summary>
        /// 商品コードで検索
        /// </summary>
        public async Task<IEnumerable<PurchaseOrderDetail>> FindByProductCodeAsync(string productCode)
        {
            const string sql = @"
                SELECT
                    発注番号 AS PoNo,
                    発注行番号 AS PoRowNo,
                    商品コード AS ProductCode,
                    商品名 AS ProductName,
                    発注単価 AS UnitPrice,
                    発注数量 AS Quantity,
                    消費税率 AS ConsumptionTaxRate,
                    入荷済数量 AS ReceivedQuantity,
                    値引金額 AS Discount,
                    指定納期 AS DueDate,
                    作成日時 AS CreatedAt,
                    作成者名 AS CreatedBy,
                    更新日時 AS UpdatedAt,
                    更新者名 AS UpdatedBy
                FROM 発注データ明細
                WHERE 商品コード = @ProductCode
                ORDER BY 発注番号, 発注行番号";

            await using var connection = new NpgsqlConnection(_connectionString);
            return await connection.QueryAsync<PurchaseOrderDetail>(sql, new { ProductCode = productCode });
        }
    }
}

File: db/fsharp/SalesManagement.Infrastructure/Repositories/PurchaseOrderDetailRepository.fs
module SalesManagement.Infrastructure.Repositories.PurchaseOrderDetailRepository

open System.Threading.Tasks
open Npgsql
open Dapper
open SalesManagement.Domain.Models

let findByPoNoAsync (connectionString: string) (poNo: string) : Task<PurchaseOrderDetail seq> =
    task {
        use connection = new NpgsqlConnection(connectionString)
        let sql = """
            SELECT
                発注番号 AS PoNo,
                発注行番号 AS PoLineNo,
                商品コード AS ProductCode,
                商品名 AS ProductName,
                発注数量 AS PoQuantity,
                入荷済数量 AS ReceivedQuantity,
                仕入単価 AS UnitPrice,
                金額 AS Amount,
                消費税 AS ConsumptionTax,
                作成日時 AS CreatedAt,
                作成者名 AS CreatedBy,
                更新日時 AS UpdatedAt,
                更新者名 AS UpdatedBy
            FROM 発注データ明細
            WHERE 発注番号 = @PoNo
            ORDER BY 発注行番号
        """
        let! results = connection.QueryAsync<PurchaseOrderDetail>(sql, {| PoNo = poNo |})
        return results
    }

let insertAsync (connectionString: string) (detail: PurchaseOrderDetail) : Task<unit> =
    task {
        use connection = new NpgsqlConnection(connectionString)
        let sql = """
            INSERT INTO 発注データ明細 (
                発注番号, 発注行番号, 商品コード, 商品名, 発注数量, 入荷済数量,
                仕入単価, 金額, 消費税,
                作成日時, 作成者名, 更新日時, 更新者名
            ) VALUES (
                @PoNo, @PoLineNo, @ProductCode, @ProductName, @PoQuantity, @ReceivedQuantity,
                @UnitPrice, @Amount, @ConsumptionTax,
                @CreatedAt, @CreatedBy, @UpdatedAt, @UpdatedBy
            )
        """
        let! _ = connection.ExecuteAsync(sql, detail)
        return ()
    }

let updateReceivedQuantityAsync (connectionString: string) (poNo: string) (lineNo: int) (receivedQuantity: int) : Task<unit> =
    task {
        use connection = new NpgsqlConnection(connectionString)
        let sql = """
            UPDATE 発注データ明細 SET
                入荷済数量 = @ReceivedQuantity,
                更新日時 = CURRENT_TIMESTAMP
            WHERE 発注番号 = @PoNo
              AND 発注行番号 = @LineNo
        """
        let! _ = connection.ExecuteAsync(sql, {| PoNo = poNo; LineNo = lineNo; ReceivedQuantity = receivedQuantity |})
        return ()
    }

let deleteByPoNoAsync (connectionString: string) (poNo: string) : Task<unit> =
    task {
        use connection = new NpgsqlConnection(connectionString)
        let sql = "DELETE FROM 発注データ明細 WHERE 発注番号 = @PoNo"
        let! _ = connection.ExecuteAsync(sql, {| PoNo = poNo |})
        return ()
    }
